src/infra/core/src/compact_service_impl.rs

```diff
@@ -0,0 +1,386 @@
+            let data_frame = ctx
+                .read_parquet(
+                    data_slice_batch_info.data_slices_batch.clone(),
+                    datafusion::execution::options::ParquetReadOptions {
+                        file_extension: "",
+                        ..Default::default()
+                    },
+                )
+                .await
+                .int_err()?;
+
+            data_frame
+                .write_parquet(
+                    data_slice_batch_info
+                        .new_file_path
+                        .as_ref()
+                        .unwrap()
+                        .to_str()
+                        .unwrap(),
+                    datafusion::dataframe::DataFrameWriteOptions::new()
+                        .with_single_file_output(true),
+                    None,
+                )
+                .await
+                .int_err()?;
+        }
+
+        Ok(())
+    }
+
+    fn create_run_compact_dir(&self) -> Result<PathBuf, CompactError> {
+        let compact_dir_path = self.run_info_dir.join("compact");
+        fs::create_dir_all(&compact_dir_path).int_err()?;
+        Ok(compact_dir_path)
+    }
+
+    async fn replace_data_files(
+        &self,
+        dataset: Arc<dyn Dataset>,
+        chain_files_info: &ChainFilesInfo,
+    ) -> Result<(Vec<Url>, Multihash), CompactError> {
+        let chain = dataset.as_metadata_chain();
+
+        let mut new_head = chain_files_info.new_head.clone();
+        let mut old_data_slices: Vec<Url> = vec![];
+
+        for data_slice_batch_info in chain_files_info.data_slice_batches.iter().rev() {
+            let add_data_params = AddDataParams {
+                prev_checkpoint: None,
+                prev_offset: data_slice_batch_info.prev_offset,
+                new_offset_interval: data_slice_batch_info.new_offset_interval.clone(),
+                new_watermark: None,
+                new_source_state: None,
+            };
+
+            let commit_result = dataset
+                .commit_add_data(
+                    add_data_params,
+                    Some(OwnedFile::new(
+                        data_slice_batch_info.new_file_path.as_ref().unwrap(),
+                    )),
+                    None,
+                    CommitOpts {
+                        block_ref: &BlockRef::Head,
+                        system_time: Some(Utc::now()),
+                        prev_block_hash: Some(Some(&new_head)),
+                        check_object_refs: false,
+                        is_reset_head: false,
+                    },
+                )
+                .await
+                .int_err()?;
+            new_head = commit_result.new_head;
+
+            // Check is new commited block equals to old one
+            // It is possible when `max-slice-size` is smaller
+            // than size of one data slice
+            if data_slice_batch_info.data_slices_batch.len() == 1 {
```

Comment:
It should not be okay. Just improve the gathering algorithm: if there is a single block that is larger than the limit, just make a batch of this single block unconditionally.

Reply:
Great point, changed